### PR TITLE
Dry up inputstream to bytesreference

### PR DESCRIPTION
**Pull request by original-brownbear**
* Dry up spots where we use the same pattern to get from an InputStream to a BytesReferences

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -645,32 +645,25 @@ public RepositoryData getRepositoryData() {
...
-                // EMPTY is safe here because RepositoryData#fromXContent calls namedObject
-                try (XContentParser parser = XContentHelper.createParser(NamedXContentRegistry.EMPTY,
-                    LoggingDeprecationHandler.INSTANCE, out.bytes(), XContentType.JSON)) {
-                    repositoryData = RepositoryData.snapshotsFromXContent(parser, indexGen);
-                } catch (NotXContentException e) {
-                    logger.warn("[{}] index blob is not valid x-content [{} bytes]", snapshotsIndexBlobName, out.bytes().length());
-                    throw e;
-                }
+            // EMPTY is safe here because RepositoryData#fromXContent calls namedObject
+            final BytesReference out = Streams.readFully(blobContainer().readBlob(snapshotsIndexBlobName));
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: I don't think we need to fully read the blob here. Instead we could pass the InputStream directly to the XContentParser and let the blob containter bufferizes the reads?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 2]
```diff
@@ -645,32 +645,25 @@ public RepositoryData getRepositoryData() {
...
             // now load the incompatible snapshot ids, if they exist
-            try (InputStream blob = blobContainer().readBlob(INCOMPATIBLE_SNAPSHOTS_BLOB)) {
-                BytesStreamOutput out = new BytesStreamOutput();
-                Streams.copy(blob, out);
-                try (XContentParser parser = XContentHelper.createParser(NamedXContentRegistry.EMPTY,
-                    LoggingDeprecationHandler.INSTANCE, out.bytes(), XContentType.JSON)) {
-                    repositoryData = repositoryData.incompatibleSnapshotsFromXContent(parser);
-                }
+            try (XContentParser parser = XContentHelper.createParser(NamedXContentRegistry.EMPTY, LoggingDeprecationHandler.INSTANCE,
+                    Streams.readFully(blobContainer().readBlob(INCOMPATIBLE_SNAPSHOTS_BLOB)), XContentType.JSON)) {
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Same here

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 3]
```diff
@@ -778,11 +771,7 @@ long latestIndexBlobId() throws IOException {
 
     // package private for testing
     long readSnapshotIndexLatestBlob() throws IOException {
-        try (InputStream blob = blobContainer().readBlob(INDEX_LATEST_BLOB)) {
-            BytesStreamOutput out = new BytesStreamOutput();
-            Streams.copy(blob, out);
-            return Numbers.bytesToLong(out.bytes().toBytesRef());
-        }
+        return Numbers.bytesToLong(Streams.readFully(blobContainer().readBlob(INDEX_LATEST_BLOB)).toBytesRef());
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: We expect this file to be very small so using `Streams.readFully()`can makes sense here.

File: test/framework/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreTestUtil.java [hunk 4]
```diff
@@ -81,14 +79,9 @@ protected void doRun() throws Exception {
...
                 final RepositoryData repositoryData;
-                try (InputStream inputStream = blobContainer.readBlob("index-" + latestGen);
-                     BytesStreamOutput out = new BytesStreamOutput()) {
-                    Streams.copy(inputStream, out);
-                    try (XContentParser parser =
-                             XContentHelper.createParser(NamedXContentRegistry.EMPTY, LoggingDeprecationHandler.INSTANCE,
-                                 out.bytes(), XContentType.JSON)) {
-                        repositoryData = RepositoryData.snapshotsFromXContent(parser, latestGen);
-                    }
+                try (XContentParser parser = XContentHelper.createParser(NamedXContentRegistry.EMPTY, LoggingDeprecationHandler.INSTANCE,
```
Comment (tlrx) [inline test/framework/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreTestUtil.java]: Same here

**Review by tlrx — COMMENTED** [review]
Nice clean up, I left some comments

**Comment by original-brownbear** [comment]
Thanks @tlrx I moved to directly deserializing in the spots you pointed out now :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 5]
```diff
@@ -645,32 +644,23 @@ public RepositoryData getRepositoryData() {
...
-                try (XContentParser parser = XContentHelper.createParser(NamedXContentRegistry.EMPTY,
-                    LoggingDeprecationHandler.INSTANCE, out.bytes(), XContentType.JSON)) {
-                    repositoryData = RepositoryData.snapshotsFromXContent(parser, indexGen);
-                } catch (NotXContentException e) {
-                    logger.warn("[{}] index blob is not valid x-content [{} bytes]", snapshotsIndexBlobName, out.bytes().length());
-                    throw e;
-                }
+            // EMPTY is safe here because RepositoryData#fromXContent calls namedObject
+            try (InputStream blob = blobContainer().readBlob(snapshotsIndexBlobName);
+                 XContentParser parser = XContentType.JSON.xContent().createParser(NamedXContentRegistry.EMPTY,
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Can you please wrap the XContentParser in a try-with-resources statement?

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: It is isn't it? :D (it's in the same try that `blob` is in)

**Review by tlrx — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 5 above)
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Pfff.... sorry for the noise

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/packaging-sample

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/2

**Comment by original-brownbear** [comment]
thanks @tlrx !

(packaging sample is just red because of the build stats cluster being down -> merging)

**Pull request merged**